spark/src/main/scala/geotrellis/spark/io/ReaderWriter.scala

```diff
@@ -37,4 +38,17 @@ abstract class FilteringLayerReader[ID, K: Boundable, ReturnType] extends LayerR
 
   def query(layerId: ID, numPartitions: Int): BoundRDDQuery[K, MetaDataType, ReturnType] =
     new BoundRDDQuery(new RDDQuery, read(layerId, _, numPartitions))
-}
\ No newline at end of file
+}
+
+abstract class LayerFormat[ID, K: Boundable, V, ReturnType] extends FilteringLayerReader[ID, K, ReturnType] with Writer[ID, ReturnType with RDD[(K, V)]] {
+  val layerReader: FilteringLayerReader[ID, K, ReturnType]
+  val layerWriter: Writer[ID, ReturnType with RDD[(K, V)]]
+
+  /**  Dirty cast, obviously MetaData types have to be equal */
+  def read(id: ID, rasterQuery: RDDQuery[K, MetaDataType], numPartitions: Int): ReturnType =
+    layerReader.read(id ,rasterQuery.asInstanceOf[RDDQuery[K, layerReader.MetaDataType]], numPartitions)
```

Comment:
We have to figure out how to take out this caste. @echeipesh says it necessitates moving the MetaData type into a type param. He's going to write up an issue on it